### PR TITLE
Build: Switch to maven-publish plugin

### PR DESCRIPTION
**Pull request by rjernst**
The build currently uses the old maven support in gradle. This commit
switches to use the newer maven-publish plugin. This will allow future
changes, for example, easily publishing to artifactory.

An additional part of this change makes publishing of build-tools part
of the normal publishing, instead of requiring a separate upload step
from within buildSrc. That also sets us up for a follow up to enable
precomit checks on the buildSrc code itself.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -17,54 +17,61 @@
...
+          Console console = System.console()
+          if (project.rootProject.hasProperty('nexusUsername') == false) {
+            project.rootProject.ext.nexusUsername = console.readLine('\nNexus username: ')
+          }
+          if (project.rootProject.hasProperty('nexusPassword') == false) {
+            project.rootProject.ext.nexusPassword = new String(console.readPassword("\nNexus password: "))
+          }
+
+          credentials {
+            username = project.rootProject.nexusUsername
```
Comment (nik9000) [inline build.gradle]: In maven this kind of thing is usually read from .m2/settings.xml. I see we didn't do that sort of thing in gradle before but is it worth considering doing it at some point? Probably not part of this PR though.

File: settings.gradle [hunk 2]
```diff
@@ -58,6 +59,8 @@ if (isEclipse) {
 
 include projects.toArray(new String[0])
 
+project(':build-tools').projectDir = new File(rootProject.projectDir, 'buildSrc')
```
Comment (nik9000) [inline settings.gradle]: Can you not use the name `buildSrc` because that is reserved by gradle somehow? I think it would be nice for the names to line up or maybe have a comment why they don't.

Comment (nik9000) [inline settings.gradle]: Can you not use the name `buildSrc` because that is reserved by gradle somehow? I think it would be nice for the names to line up or maybe have a comment why they don't.

**Comment by nik9000** [comment]
Left two questions but they shouldn't block merging. LGTM.

File: build.gradle (hunk 1 above)
Comment (rjernst) [inline build.gradle]: These are already being read as gradle properties (both now and before). You can either add them on the command line with eg `gradle -PnexusUsername=myuser publish`, or you can add `nexusUsername=myuser` in your home gradle properties eg `~/.gradle/gradle.properties`.

File: settings.gradle (hunk 2 above)
Comment (rjernst) [inline settings.gradle]: I don't want to use the name buildSrc for the artifact because I think it is confusing. It is only called buildSrc in the project because this is required by gradle.
